### PR TITLE
TensorGraph

### PR DESCRIPTION
**Pull request by lilleswing**
Creates a new framework for constructing models.  New Framework allows for arbitrary graphs to be created.

To see how it feels read through MINST.py in /examples/tensorflow and 
deepchem/models/tensorgraph/models/robust_multitask.py
which is a re-implementation of robust multitask without dropout or regularization yet.

Still TODO
* More Tests, want 100% test coverage on TensorGraph
* Documentation on layers
..* Layer documentation is the most important, I also want to add more assertions for parent layers.
* Finish the tutorials layed out in https://github.com/lilleswing/deepchem/blob/tensorgraph-cr/docs/using_tensorgraph.md
..* TensorGraph for Bypass MultiTaskClassification
..* Using Tensorboard with TensorGraph
..* Experimenting with new architectures with TensorGraph
..* Adding complexity while fighting overfitting
..* Transfer Learning
* Rewrite all TF based models in benchmark.py to use TensorGraph style.

I want to get the body in, in case anyone wants to play with it, and so code doesn't drift too far despite still having the long TODO list.

### CODE REVIEW

File: deepchem/models/tensorgraph/layers.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+
+class Input(Layer):
+
+  def __init__(self, shape, **kwargs):
+    self.t_shape = shape
+    super().__init__(**kwargs)
+
+  def __call__(self, *parents):
+    self.out_tensor = tf.placeholder(tf.float32, shape=self.t_shape)
```
Comment (peastman) [inline deepchem/models/tensorgraph/layers.py]: Since these are new classes, you might want to design them from the beginning to be able to work with queues.  That means not hardcoding placeholders like this.  During evaluation, the inputs to the graph are indeed placeholders.  But during training, the placeholders are the inputs to the queue, and the outputs from the queue become the inputs to the graph.  So it needs to be able to do either one, depending on which purpose you're building the graph for.  See #488.

File: deepchem/models/tensorgraph/tensor_graph.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+    """
+    if len(self.features) != 1:
+      raise ValueError("Only allow one input set of features")
+    features = self.features[0]
+    if not self.built:
+      self.build()
+    with self.tensor_objects['Graph'].as_default():
+      saver = tf.train.Saver()
+      with tf.Session() as sess:
+        saver.restore(sess, self.last_checkpoint)
```
Comment (rbharath) [inline deepchem/models/tensorgraph/tensor_graph.py]: Restoring on every batch is going to get pretty slow. I'd recommend also directly overriding the `predict()` method so that only one restore is needed for large `predict()` operations.

File: deepchem/models/tensorgraph/tensor_graph.py [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+
+    # add out_tensor back to everyone
+    if must_restore:
+      for index, node in enumerate(self.topsort()):
+        node_layer = self.layers[node]
+        node_layer.out_tensor = out_tensors[index]
+      self.built = True
+    self.tensor_objects = tensor_objects
+
+  def _get_tf(self, obj):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/tensor_graph.py]: I'm a little uncertain about having an internal parameter dictionary. I can see why it's convenient, but there's also a risk of having it balloon with unwanted objects in subclasses. This is more a soft feeling than a hard request though. Perhaps more docs could help control the dictionary.

File: docs/MNIST.md [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+tg = dc.models.TensorGraph(tensorboard=True, model_dir='/tmp/mnist')
+feature = Input(shape=(None, 784))
+tg.add_layer(feature)
+tg.add_feature(feature)
+
+# Images are square 28x28 (batch, height, width, channel)
+make_image = Reshape(shape=(-1, 28, 28, 1))
+tg.add_layer(make_image, parents=[feature])
+
+class Conv2d(Layer):
```
Comment (rbharath) [inline docs/MNIST.md]: We should probably move `Conv2d` directly into DeepChem, especially since we want to eventually support more visual/microscopy datasets moving forward.

File: docs/using_tensorgraph.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
+# Using TensorGraph
```
Comment (rbharath) [inline docs/using_tensorgraph.md]: The introductory material here looks good!

We should add a citation to Keras (or just credit them for the design informally) since we borrowed a lot of the stylistic form from them. It might also be useful to have a "how is this not keras" section that explains the differing semantics of `Layer` objects. This is more for advanced users, but I anticipate we'll get at least a few people who have tried Keras before trying the new Tensorgraphs.

**Comment by rbharath** [comment]
@lilleswing Design looks good at a high level :+1:. I've added some review comments at a few places in the code, but I'm pretty happy with the overall design.

As @peastman points out, it's worth figuring out the story with queues and Tensorgraphs though. Probably something easiest to discuss offline.

File: deepchem/models/tensorgraph/tensor_graph.py (hunk 3 above)
Comment (lilleswing) [inline deepchem/models/tensorgraph/tensor_graph.py]: Yeah I think its solved via documentation.  The only reason it is necessary is to temporarily set it to None while pickling the object.  An other option would be a container class which lazily instantiates these objects.

File: docs/MNIST.md (hunk 4 above)
Comment (lilleswing) [inline docs/MNIST.md]: Yeah I can move Conv2d inside and split maxpool out as well.

File: deepchem/models/tensorgraph/models/robust_multitask.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  model_dir
+
+  Returns
+  -------
+
+  """
+  g = MultiTaskTensorGraph(model_dir=model_dir)
+  in_layer = Input(shape=(None, n_features), name="FEATURE")
+  g.add_layer(in_layer)
+  g.add_feature(in_layer)
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/models/robust_multitask.py]: sure

Comment (rbharath) [inline deepchem/models/tensorgraph/models/robust_multitask.py]: Great. Might be useful to do the same for `add_label()` and `set_loss()` as well

**Comment by lilleswing** [comment]
@rbharath @peastman I hooked up a queue to be used for fitting models automatically based on the features, labels, and weights.

We create new corollary Input objects whose name is "%s_pre_q" for each input.  These are inputs that we feed into a queue, and the user defined inputs now are tf.placeholder_with_default(q.out).  

More work needs to be done for closing the queue correctly, and persisting (The documentation is really unclear with why and when you have to close the queue, and once you close it you can't reopen it).  However this is basically the approach I plan on taking for using a Queue for async feeding data into the network.

**Comment by peastman** [comment]
The way I handle it in #488 is that a worker thread is feeding data into the queue, and when it's all done it closes the queue.  The main thread is then running the training operations, which pulls data off the queue.  When it reaches the end of the data, `sess.run()` throws an `OutOfRangeError`, which is how it knows it's done.  This avoids the complications of both threads needing to know in advance exactly how many batches of training data there will be.  If the queue weren't closed, the main thread would hang indefinitely waiting for more data instead of exiting immediately.

File: deepchem/models/tensorgraph/models/robust_multitask.py (hunk 6 above)
Comment (lilleswing) [inline deepchem/models/tensorgraph/models/robust_multitask.py]: I added it to add_label, but I don't want to add parents=[] as a param to set_loss which is required because a loss has to be wired up.

**Comment by rbharath** [comment]
Based on offline discussions, we should be good to merge this PR in once tests pass. We will incorporate the TODOs and fixes in following PRs.

@miaecle Would you mind reading through this PR (even if it's been merged) and commenting on whether the design is adequate to implement the DTNNs and the DAG models? We can address any issues you find in future PRs.

**Comment by lilleswing** [comment]
@rbharath this is good to go now.

**Comment by rbharath** [comment]
LGTM.
